src/WebJobs.Extensions.DurableTask/LifeCycleNotificationHelper.cs

```diff
@@ -0,0 +1,243 @@
+                httpClient = new HttpClient();
+                INameResolver nameResolver = extensionConfigContext.Config.GetService<INameResolver>();
+                this.eventGridKeyValue = nameResolver.Resolve(config.EventGridKeySettingName);
+                httpClient.DefaultRequestHeaders.Add("aeg-sas-key", this.eventGridKeyValue);
+            }
+        }
+
+        private bool UseTrace { get; }
+
+        public void SetHttpMessageHandler(HttpMessageHandler handler)
+        {
+            httpClient = new HttpClient(handler);
+            httpClient.DefaultRequestHeaders.Add("aeg-sas-key", this.eventGridKeyValue);
+        }
+
+        private async Task SendNotificationAsync(
+            EventGridEvent[] eventGridEventArray,
+            string hubName,
+            string functionName,
+            string version,
+            string instanceId,
+            string reason,
+            FunctionType functionType,
+            bool isReplay,
+            FunctionState functionState)
+        {
+            var json = JsonConvert.SerializeObject(eventGridEventArray);
+            var content = new StringContent(json, Encoding.UTF8, "application/json");
+
+            // Details about the Event Grid REST API
+            // https://docs.microsoft.com/en-us/rest/api/eventgrid/
+            var result = await httpClient.PostAsync(this.config.EventGridTopicEndpoint, content);
+
+            if (!result.IsSuccessStatusCode)
+            {
+                var appName = Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME") ?? string.Empty;
+                var slotName = Environment.GetEnvironmentVariable("WEBSITE_SLOT_NAME") ?? string.Empty;
+                var extensionVersion = FileVersionInfo.GetVersionInfo(typeof(DurableTaskExtension).Assembly.Location).FileVersion;
+
+                this.config.TraceHelper.SendMessageFailed(
+                    hubName,
+                    functionName,
+                    functionState,
+                    version,
+                    instanceId,
+                    result.StatusCode,
+                    result.ReasonPhrase,
+                    reason);
+            }
+        }
+
+        public async Task OrchestratorStartingAsync(
+            string hubName,
+            string functionName,
+            string version,
+            string instanceId,
+            FunctionType functionType,
+            bool isReplay)
+        {
+            if (!this.UseTrace)
+            {
+                return;
+            }
+
+            EventGridEvent[] sendObject = this.CreateEventGridEvent(
+                hubName,
+                functionName,
+                version,
+                instanceId,
+                "",
+                OrchestrationRuntimeStatus.Running);
+            await this.SendNotificationAsync(sendObject, hubName, functionName, version, instanceId, "", FunctionType.Orchestrator, isReplay, FunctionState.Started);
```

Comment:
All the async methods below only use the `await` keyword as the very last statement. Because of that, I suggest we not use async/await compiler feature and instead return the task from `this.SendNotificationAsync` directly. This has the benefit of simplifying exception debugging and should result in a slight performance improvement since the code does not need to go through the compiler-generated async state machine.

For example:
```csharp
public Task OrchestratorCompletedAsync(...)
{
    // ...
    return this.SendNotificationAsync(...);
}
```